### PR TITLE
SONARJS-822 Rule: Jump statements should not be used unconditionally

### PR DESCRIPTION
**Pull request by pynicolas**

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnconditionalJumpStatementCheck.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+@Rule(key = "S1751")
+public class UnconditionalJumpStatementCheck extends SubscriptionVisitorCheck {
+
+  private static final Kind[] LOOP_KINDS = {
+    Kind.WHILE_STATEMENT,
+    Kind.DO_WHILE_STATEMENT,
+    Kind.FOR_STATEMENT,
+    Kind.FOR_OF_STATEMENT
+  };
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnconditionalJumpStatementCheck.java]: @pynicolas formatting: move this on previous line

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnconditionalJumpStatementCheck.java [hunk 2]
```diff
@@ -79,12 +80,20 @@ private static boolean canExecuteMoreThanOnce(IterationStatementTree loopTree) {
...
-    for (CfgBlock loopPredecessor : block.predecessors()) {
-      StatementTree loopBody = loopTree.statement();
+  private static boolean hasPredecessorInsideLoopBody(CfgBranchingBlock conditionBlock, IterationStatementTree loopTree) {
+    for (CfgBlock loopPredecessor : conditionBlock.predecessors()) {
       List<Tree> predecessorElements = loopPredecessor.elements();
-      Tree lastElement = predecessorElements.get(predecessorElements.size() - 1);
-      if (isDescendant(lastElement, loopBody)) {
+      Tree predecessorLastElement = predecessorElements.get(predecessorElements.size() - 1);
+
+      if (loopTree.is(Kind.FOR_STATEMENT)) {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnconditionalJumpStatementCheck.java]: @pynicolas We can remove this "hack" if improve CFG and check not last but first element

File: javascript-checks/src/test/resources/checks/UnconditionalJumpStatement.js [hunk 3]
```diff
@@ -81,3 +81,16 @@ while(foo()) {
...
+    continue;
+  }
+  baz();
+  break; // compliant
+}
+
+for (i = 0; foo();) {
+  baz();
+  break; // Noncompliant
+}
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/UnconditionalJumpStatement.js]: @pynicolas can we have a case for `for` loop with noncompliant `continue`?

File: javascript-checks/src/test/resources/checks/UnconditionalJumpStatement.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+for(p in obj) {
+  bar();
+  break; // Compliant: often used to check whether an object is "empty"
+}
+
+for(p of obj) {
+  bar();
+  break; // Noncompliant
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/UnconditionalJumpStatement.js]: @pynicolas what about raising issue for 
```
for (p in obj) { foo(); continue; }
```
?

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnconditionalJumpStatementCheck.java (hunk 2 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnconditionalJumpStatementCheck.java]: @vilchik-elena We could improve the CFG, but it's not easy: I don't think it's worth it.

File: javascript-checks/src/test/resources/checks/UnconditionalJumpStatement.js (hunk 4 above)
Comment (pynicolas) [inline javascript-checks/src/test/resources/checks/UnconditionalJumpStatement.js]: @vilchik-elena Done in 977ed8c

File: javascript-checks/src/main/java/org/sonar/javascript/checks/UnconditionalJumpStatementCheck.java (hunk 1 above)
Comment (pynicolas) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/UnconditionalJumpStatementCheck.java]: @vilchik-elena Changed in e3226ef

File: javascript-checks/src/test/resources/checks/UnconditionalJumpStatement.js (hunk 3 above)
Comment (pynicolas) [inline javascript-checks/src/test/resources/checks/UnconditionalJumpStatement.js]: @vilchik-elena Done in a22d781

**Pull request merged**